src/test_typing_extensions.py

```diff
@@ -7173,6 +7173,82 @@ def test_attributes(self):
         self.assertEqual(Variadic.__type_params__, (Ts,))
         self.assertEqual(Variadic.__parameters__, tuple(iter(Ts)))
 
+        P = ParamSpec('P')
+        CallableP = TypeAliasType("CallableP", Callable[P, Any], type_params=(P, ))
+        self.assertEqual(CallableP.__name__, "CallableP")
+        self.assertEqual(CallableP.__value__,  Callable[P, Any])
+        self.assertEqual(CallableP.__type_params__, (P,))
+        self.assertEqual(CallableP.__parameters__, (P,))
+
+    def test_alias_types_and_substitutions(self):
+        T = TypeVar('T')
+        T2 = TypeVar('T2')
+        T_default = TypeVar("T_default", default=int)
+        Ts = TypeVarTuple("Ts")
+        P = ParamSpec('P')
+
+        test_argument_cases = {
+            # arguments : expected parameters
+            int : (),
+            ... : (),
+            T2 : (T2,),
+            Union[int, List[T2]] : (T2,),
+            Tuple[int, str] : (),
+            Tuple[T, T_default, T2] : (T, T_default, T2),
+            Tuple[Unpack[Ts]] : (Ts,),
+            Callable[[Unpack[Ts]], T2] : (Ts, T2),
+            Callable[P, T2] : (P, T2),
+            Callable[Concatenate[T2,  P], T_default] : (T2, P, T_default),
+            TypeAliasType("NestedAlias", List[T], type_params=(T,))[T2] : (T2,),
+            Unpack[Ts] : (Ts,),
+            Unpack[Tuple[int, T2]] : (T2,),
+            Concatenate[int,  P] : (P,),
+        }
+        test_argument_cases_311_plus = {
+            Ts : (Ts,),  # invalid case
+        }
+        test_argument_cases.update(test_argument_cases_311_plus)
+
+        test_alias_cases = [
+            # Simple cases
+            TypeAliasType("ListT", List[T], type_params=(T,)),
+            TypeAliasType("UnionT", Union[int, List[T]], type_params=(T,)),
+            # Value has no parameter but in type_param
+            TypeAliasType("ValueWithoutT", int, type_params=(T,)),
+            # Callable
+            TypeAliasType("CallableP", Callable[P, Any], type_params=(P, )),
+            TypeAliasType("CallableT", Callable[..., T], type_params=(T, )),
+            TypeAliasType("CallableTs", Callable[[Unpack[Ts]], Any], type_params=(Ts, )),
+            # TypeVarTuple
+            TypeAliasType("Variadic", Tuple[int, Unpack[Ts]], type_params=(Ts,)),
+            # TypeVar with default
+            TypeAliasType("TupleT_default", Tuple[T_default, T], type_params=(T, T_default)),
+            TypeAliasType("CallableT_default", Callable[[T], T_default], type_params=(T, T_default)),
+            # default order reversed
+            TypeAliasType("TupleT_default_reversed", Tuple[T_default, T], type_params=(T_default, T)),
```

Comment:
Hm, this should be invalid. It's a SyntaxError with the native version:

```
>>> type X[T=int, U] = (T, U)
  File "<python-input-0>", line 1
    type X[T=int, U] = (T, U)
                  ^
SyntaxError: non-default type parameter 'U' follows default type parameter
```

We should also check against this in TypeAliasType in CPython, and then backport to typing-extensions.

Reply:
I kind of expected it it be invalid. There is ... a thought... of an issue in my head about default values, but I am not yet sure if it is based on a mistake of mine, if it should be a dicussion first or if it should be reported as a bug (and to which repo).

I think I should move this test to another region (or out of this PR) to keep better track of it. Any suggestions on this?

Reply:
I would:

- Report a bug to CPython about this. The behavior when using the TypeAliasType constructor should match that of the type statement, so it should be an error to have a type parameter without a default following one with a default.
- Address it in CPython and then backport to typing-extensions
- Remove any tests from this PR that rely on behavior that should be an error because of this bug.

Reply:
Removed the tests and created https://github.com/python/cpython/issues/124787